### PR TITLE
intel_adsp: ace: secondary core context save and restore

### PR DESCRIPTION
**Pull request by tmleman**
Reusing primary core context save/restore flow for purpose of secondary core D0 -> D3 -> D0 transitions.

FW on secondary core when preparing for D3 will save its context. When core is re-enabled we use _**dsp_restore_vector**_ as the FW entry point.

### CODE REVIEW

File: soc/xtensa/intel_adsp/ace/power.c [hunk 1]
```diff
@@ -176,8 +188,11 @@ __asm__(".align 4\n\t"
 	"  wsr   a0, WINDOWBASE\n\t"
 	"  rsync\n\t"
 	"  movi  sp, d0i3_stack\n\t"
-	"  movi a2, 0x1000\n\t"
-	"  add sp, sp, a2\n\t"
+	"  rsr   a2, PRID\n\t"
+	"  movi  a3, 0x400\n\t"
+	"  mull  a2, a2, a3\n\t"
+	"  add   a2, a2, a3\n\t"
+	"  add   sp, sp, a2\n\t"
```
Comment (andyross) [inline soc/xtensa/intel_adsp/ace/power.c:195]: What's going on here?  You're computing ((PRID + 1) * 1024) and popping that many bytes off the stack?  It was weird enough before where there seems to have been a 4k buffer being adjusted, but I don't understand at all why the stack adjustment needs to be CPU-specific.  This is the CPU-local interrupt stack, isn't it?

File: soc/xtensa/intel_adsp/ace/power.c [hunk 2]
```diff
@@ -323,7 +334,7 @@ __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 			imr_layout->imr_state.header.imr_ram_storage = NULL;
 		}
 #endif /* CONFIG_ADSP_IMR_CONTEXT_SAVE */
-
+		z_mark_thread_as_not_suspended(z_current_get());
```
Comment (andyross) [inline soc/xtensa/intel_adsp/ace/power.c:337]: This isn't right.  The "suspended" state means that a thread is neither running currently nor in the run queue[1].  By definition, an idle thread can never be blocked, and by implementation choice is never added to the run queue.  It just doesn't make sense here.

File: soc/xtensa/intel_adsp/ace/power.c [hunk 3]
```diff
@@ -222,9 +237,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		z_xt_ints_off(0xffffffff);
 		core_desc[cpu].bctl = DSPCS.bootctl[cpu].bctl;
 		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPCG;
-		soc_cpus_active[cpu] = false;
-		z_xtensa_cache_flush_inv_all();
 		if (cpu == 0) {
+			soc_cpus_active[cpu] = false;
```
Comment (andyross) [inline soc/xtensa/intel_adsp/ace/power.c:241]: Won't this now fail to update soc_cpus_active[] for secondare cores?

**Review by andyross — CHANGES_REQUESTED** [review]
Can we clean this up?  It seems like you needed a way to tell whether or not a core was being started up after initial boot, and decided to abuse the kernel thread state for the idle thread?  That's... no, that's just not right.  Those are different subsystems.  If the arch/soc code needs to track state for its own purposes, it needs to do it on its own and not repurpose other code from other areas.

FWIW: looking at that soc_cpus_active[] array, I'm guessing that's the right kind of abstraction to be using (though it seems to have been broken by this PR?).

File: soc/xtensa/intel_adsp/ace/power.c (hunk 1 above)
Comment (andyross) [reply soc/xtensa/intel_adsp/ace/power.c:195]: OK, now I get it.  This is just computing the end of the relevant subarray, and doing the computation at runtime because the __asm__ block is global and so can't feed C expressions via constraints.  That's annoying, I've been there.  But can you clean this up so you aren't hardcoding 0x400 == CONFIG_IDLE_STACK_SIZE, or at the very least put a comment here explaining what's going on?

Still, I don't get the point here.  Why are you wasting 4k (!!) of precious HPSRAM for per-core power stacks, when there's a perfectly good and completely unused (because interrupts are masked, right?  right?!) per-core interrupt stack already.

Please find a way to unify this with the interrupt stack.

File: soc/xtensa/intel_adsp/ace/power.c (hunk 3 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c:241]: No. For secondary core its done inside of _power_gate_entry_ function.

**Comment by tmleman** [comment]
@andyross please take a look at the updated versions now.

**Review by jxstelter — DISMISSED** [review]

File: soc/xtensa/intel_adsp/ace/power.c [hunk 4]
```diff
@@ -163,9 +166,12 @@ __asm__(".align 4\n\t"
 	"  wsr   a1, WINDOWSTART\n\t"
 	"  wsr   a0, WINDOWBASE\n\t"
 	"  rsync\n\t"
-	"  movi  sp, d0i3_stack\n\t"
-	"  movi a2, 0x1000\n\t"
-	"  add sp, sp, a2\n\t"
+	"  movi  sp, z_interrupt_stacks\n\t"
+	"  rsr   a2, PRID\n\t"
+	"  movi  a3, 0x800\n\t"
```
Comment (dcpleung) [inline soc/xtensa/intel_adsp/ace/power.c:183]: Could you try `"movi a3, " CONFIG_ISR_STACK_SIZE "\n\t"`? This should embed the value and you can skip the BUILD_ASSERT above.

**Review by aborisovich — DISMISSED** [review]

File: soc/xtensa/intel_adsp/ace/power.c (hunk 4 above)
Comment (softwarecki) [reply soc/xtensa/intel_adsp/ace/power.c:183]: `"movi a3, " STRINGIFY(CONFIG_ISR_STACK_SIZE)  "\n\t"` should work ;)

File: soc/xtensa/intel_adsp/ace/multiprocessing.c [hunk 5]
```diff
@@ -7,6 +7,8 @@
 #include <zephyr/kernel.h>
 #include <zephyr/sys/check.h>
 #include <zephyr/arch/cpu.h>
+#include <zephyr/pm/pm.h>
+#include <ksched.h>
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/multiprocessing.c:11]: why are you including `ksched.h` here ?

File: soc/xtensa/intel_adsp/ace/multiprocessing.c [hunk 6]
```diff
@@ -82,9 +84,21 @@ void soc_start_core(int cpu_num)
...
 		/* Initialize the ROM jump address */
 		uint32_t *rom_jump_vector = (uint32_t *) ROM_JUMP_ADDR;
-		*rom_jump_vector = (uint32_t) z_soc_mp_asm_entry;
+
+		/* We need to find out what type of booting is taking place here. Secondary cores
+		 * can be disabled and enabled multiple times during runtime. During kernel
+		 * initialization, the next pm state is set to ACTIVE. This way we can determine
+		 * whether the core is being turned on again or for the first time.
+		 */
+		if (pm_state_next_get(cpu_num)->state == PM_STATE_ACTIVE) {
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/multiprocessing.c:96]: PM is optional, one can build without it. If you build wihout PM, `pm_state_next_get()` will return a NULL pointer. This change can cause a null pointer dereference. Please account this possibility here.

File: soc/xtensa/intel_adsp/ace/power.c [hunk 7]
```diff
@@ -163,9 +178,12 @@ __asm__(".align 4\n\t"
 	"  wsr   a1, WINDOWSTART\n\t"
 	"  wsr   a0, WINDOWBASE\n\t"
 	"  rsync\n\t"
-	"  movi  sp, d0i3_stack\n\t"
-	"  movi a2, 0x1000\n\t"
-	"  add sp, sp, a2\n\t"
+	"  movi  sp, z_interrupt_stacks\n\t"
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:181]: isn't `a1` the `sp` ? It is just confuse to me mixing nomenclature

File: soc/xtensa/intel_adsp/ace/power.c [hunk 8]
```diff
@@ -134,27 +134,42 @@ void dsp_restore_vector(void);
 
 void power_gate_entry(uint32_t core_id)
 {
-	struct lpsram_header *lpsheader =
-		(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
-
 	xthal_window_spill();
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:137]: I particularly don't like using HAL API when we have equivalent implemented. We could use our API the limitation is that we need save `a0` and do `call0` in assembler.

File: soc/xtensa/intel_adsp/ace/power.c [hunk 9]
```diff
@@ -210,9 +228,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		z_xt_ints_off(0xffffffff);
 		core_desc[cpu].bctl = DSPCS.bootctl[cpu].bctl;
 		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPCG;
-		soc_cpus_active[cpu] = false;
-		z_xtensa_cache_flush_inv_all();
 		if (cpu == 0) {
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:231]: Why did you move this if ? Don't you need `soc_cpus_active[cpu] = false;` for cores != 0 too ?

**Review by ceolin — CHANGES_REQUESTED** [review]
Giving a  -1 to avoid it be merged why there are open questions

File: soc/xtensa/intel_adsp/ace/multiprocessing.c (hunk 5 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/multiprocessing.c]: Thanks for catching this. Those are leftovers from the previous version. I forgot to delete while editing.

File: soc/xtensa/intel_adsp/ace/power.c (hunk 7 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c:176]: Yes, they are the same. But I don't see how it's confusing. In this part of code I'm calculating correct stack pointer.

File: soc/xtensa/intel_adsp/ace/power.c (hunk 8 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c:137]: Could you explain more precisely what you mean?

File: soc/xtensa/intel_adsp/ace/power.c (hunk 9 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c:226]: For secondary core its done in _**power_gate_entry**_ function.

**Comment by ceolin** [comment]
> Yes, they are the same. But I don't see how it's confusing. In this part of code I'm calculating correct stack pointer. 

I know, but mixing names is confusing IMHO. I saw only one more place using `sp` for `a1`, everywhere else we are just using `a1`. This may be confuse for someone that is not super used to Xtensa.

File: soc/xtensa/intel_adsp/ace/power.c (hunk 8 above)
Comment (ceolin) [reply soc/xtensa/intel_adsp/ace/power.c:136]: I'm just saying that I prefer using Zephyr's native API (when avaialable) than the HAL API. That's said the API to spill window register on Zephyr has constraints different from the one in the HAL and I was explaining that.

Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c:136]: I prefer to keep it as it is. This PR is resolving issues with secondary cores power flows and I don't wanna do too much refactoring in it.

Comment (ceolin) [reply soc/xtensa/intel_adsp/ace/power.c:136]: that is reasonabe

File: soc/xtensa/intel_adsp/ace/power.c [hunk 10]
```diff
@@ -210,9 +222,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
 		z_xt_ints_off(0xffffffff);
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:222]: Why do you need this here ? I'm looking these masks and I don't understand. This function enters will interruptions masked XCHAL_NMILEVEL. If you don't need to do it, you could move `core_desc[cpu].intenable = XTENSA_RSR("INTENABLE");` to `_save_core_context` and avoid unecessary instructions and duplicated code.

File: soc/xtensa/intel_adsp/ace/power.c [hunk 11]
```diff
@@ -210,9 +222,9 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
...
 		core_desc[cpu].bctl = DSPCS.bootctl[cpu].bctl;
 		DSPCS.bootctl[cpu].bctl &= ~DSPBR_BCTL_WAITIPCG;
-		soc_cpus_active[cpu] = false;
-		z_xtensa_cache_flush_inv_all();
 		if (cpu == 0) {
+			soc_cpus_active[cpu] = false;
+			z_xtensa_cache_flush_inv_all();
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:227]: couldn't this be done inside `save_core_context` too ? Because you are doing this big cache invalidation and then do other times in this function for smaller memory chunks. If you put it in the end you could do this flush/invalidation only once.

**Comment by ceolin** [comment]
Please take a look into my comments. They are not things added now, but since there is you are changing some parts of this code they make sense. This logic is getting overcomplicated and confusing IMHO

File: soc/xtensa/intel_adsp/ace/power.c (hunk 10 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c:222]: For primary core _save_core_context is done only when CONFIG_ADSP_IMR_CONTEXT_SAVE is enabled.

**Comment by tmleman** [comment]
> Please take a look into my comments. They are not things added now, but since there is you are changing some parts of this code they make sense. This logic is getting overcomplicated and confusing IMHO

@ceolin I have addressed your comments in some way, please see if this version suits you.

File: soc/xtensa/intel_adsp/ace/power.c (hunk 11 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c]: In case of a secondary core I want to change this value only when everything is ready to turn it off.

File: soc/xtensa/intel_adsp/ace/power.c [hunk 12]
```diff
@@ -134,27 +133,37 @@ void dsp_restore_vector(void);
 
 void power_gate_entry(uint32_t core_id)
 {
-	struct lpsram_header *lpsheader =
-		(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
-
 	xthal_window_spill();
+	z_xtensa_cache_flush_inv_all();
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:137]: why don't you do this in the end of this function and avoid the following flushes ?

File: soc/xtensa/intel_adsp/ace/power.c [hunk 13]
```diff
@@ -134,27 +133,37 @@ void dsp_restore_vector(void);
...
 	_save_core_context(core_id);
-	lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
-	lpsheader->lp_restore_vector = &dsp_restore_vector;
+	if (core_id == 0) {
+		struct lpsram_header *lpsheader =
+			(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
+
+		lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
+		lpsheader->lp_restore_vector = &dsp_restore_vector;
+		z_xtensa_cache_flush(lpsheader, sizeof(struct lpsram_header));
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:145]: here

File: soc/xtensa/intel_adsp/ace/power.c [hunk 14]
```diff
@@ -134,27 +133,37 @@ void dsp_restore_vector(void);
...
+		/* Re-enabling interrupts for core 0 because someone has to wake-up us
+		 * from power gaiting.
+		 */
+		z_xt_ints_on(ALL_USED_INT_LEVELS_MASK);
+	}
+
 	soc_cpus_active[core_id] = false;
-	z_xtensa_cache_flush_inv_all();
-	z_xt_ints_on(ALL_USED_INT_LEVELS_MASK);
+	z_xtensa_cache_flush(soc_cpus_active, sizeof(soc_cpus_active));
```
Comment (ceolin) [inline soc/xtensa/intel_adsp/ace/power.c:153]: and here

File: soc/xtensa/intel_adsp/ace/power.c (hunk 12 above)
Comment (tmleman) [reply soc/xtensa/intel_adsp/ace/power.c:137]: I need to do cache flush/invalidation for all after window spill and before context saving. It will not work otherwise.

**Comment by tmleman** [comment]
@andyross friendly reminder

**Review by aborisovich — DISMISSED** [review]

**Comment by tmleman** [comment]
Adding DNM label because of https://github.com/thesofproject/sof/issues/7433. After this PR is merged zephyr update on SOF side will require additional changes. This would make later bisect more difficult in case of any new problems.

**Review by ceolin — DISMISSED** [review]
I'm still a little bit skeptical about the need of all flush/invalidate cache but you know better than me about it here

**Comment by ceolin** [comment]
I'm still a little bit skeptical about the need of all flush/invalidate cache but you know better than me about it here

**Comment by tmleman** [comment]
Rebase after merge of  https://github.com/zephyrproject-rtos/zephyr/pull/50136

**Review by andyross — APPROVED** [review]
Saw this come by again.  Looks great to me, my concerns earlier were all addressed.

**Review by ceolin — APPROVED** [review]

**Review by softwarecki — APPROVED** [review]

**Review by jxstelter — APPROVED** [review]
